### PR TITLE
PLT-8282: RTK Query + Redux persist refactor

### PR DESCRIPTION
**Pull request by EehMauro**
@coderabbitai: ignore

### CODE REVIEW

**Review by amnambiar — CHANGES_REQUESTED** [review]
- Upon regression, between Testing and History and running tests, and a point of inactivity, received this error in console-
image

After this it threw multiple undefined console errors and the tool went unresponsive.

File: src/api/api.ts [hunk 1]
```diff
@@ -1,57 +0,0 @@
...
-    "Content-type": "application/json",
-    Accept: "application/json",
-  },
-});
-
-export const postData = axios.create({
-  baseURL: BASE_URL,
-  headers: {
-    "Content-type": "text/plain;charset=utf-8",
-    Accept: "text/plain;charset=utf-8",
```
Comment (amnambiar) [inline src/api/api.ts:18]: This needs to be migrated over. 

TO BE used in case of sending commitHash in payload; `postData.post("/run", checkout)`

File: src/api/api.ts [hunk 2]
```diff
@@ -1,57 +0,0 @@
...
-  onRequestError
-);
-
-postData.interceptors.request.use(
-  (config: AxiosRequestConfig) => onRequest(config),
-  onRequestError
-);
-
-getRepoAccess.interceptors.request.use(
-  (config: AxiosRequestConfig) => onRepoAccessRequest(config),
```
Comment (amnambiar) [inline src/api/api.ts:55]: Also needs TO BE used, to intercept the request , to make sure accessToken is set in the Authorization header, instead of authToken. To ensure connectivity to a private repository.

Comment (amnambiar) [reply src/api/api.ts:55]: These both in this file needs to be migrated as well as used in the app.

File: src/api/onRequest.ts [hunk 3]
```diff
@@ -1,40 +0,0 @@
...
-  return config;
-}
-
-export function onRequestError(error: AxiosError) {
-  if (localStorage.getItem(LocalStorageKeys.isLoggedIn) !== "true") {
-    localStorage.clear();
-    window.location.reload();
-  }
-  return Promise.reject(error);
-}
```
Comment (amnambiar) [inline src/api/onRequest.ts:25]: Also, this - API error handling ,, Or will we handle this in https://input-output.atlassian.net/browse/PLT-8283

File: src/components/CertificationMetadataForm/index.tsx [hunk 4]
```diff
@@ -68,11 +68,7 @@ const CertificationMetadataForm = (props: Props) => {
         ...(!props.isReviewCertification ? {certificationLevel: parseInt(certificationLevel)} : {}),
         ...(!props.isReviewCertification ? {report: report.map(report => report.value)} : {}),
         ...(!props.isReviewCertification ? {subject: subject} : {})
-      },
-      profile: props.isReviewCertification ? { 
-        ...profile, 
-        dapp: subject ? { ...profile?.dapp, subject } : (profile ? profile.dapp : null)
-      } as any : null
```
Comment (amnambiar) [inline src/components/CertificationMetadataForm/index.tsx]: Why is this removed? 

I have some doubts over the master-pull made in this branch - 

1. Make sure PR #30 changes/fixes are retained. I can see https://github.com/input-output-hk/dapps-certification-web/commit/26c454fdd00cfe64b834ec501d2c97262c5064bb#diff-0ac7528762c66094096f63915c8293d4710767d1dedeebf500f086169deb5a06R72-R75 being added there.

File: src/components/CreateCertificate/CreateCertificate.tsx [hunk 5]
```diff
@@ -1,237 +1,195 @@
...
+          onClose={() => setShowError("")}
+        >
+          {showError}
+        </Alert>
+      </Snackbar>
+    </>
+  );
 }
 
 export default CreateCertificate;
```
Comment (amnambiar) [inline src/components/CreateCertificate/CreateCertificate.tsx]: Can you try to redo the changes in this file, so that only they are highlighted, rather than edits appearing in all lines here. For easiness of maintenance, and to track(`git blame`) while figuring out edits.

File: src/components/LogsView/LogsView.css [hunk 6]
```diff
@@ -6,9 +6,7 @@
   .logContainer {
     margin: 10px;
     text-align: end;
-  }
-  @media (max-width: 1199px) {
-    .logContainer {
```
Comment (amnambiar) [inline src/components/LogsView/LogsView.css:11]: This needs to be reintroduced. To explicitly specify the selector in media-query rule.

After removing sass, we can no more nest properties; unless we introduce support to nesting. It threw many build warnings and that is why I had removed them. They were just warnings, and not errors, but still.

File: src/compositions/Timeline/Timeline.tsx [hunk 7]
```diff
@@ -13,7 +13,8 @@ const Timeline = (props: any) => {
     <div className="statusTimeline" data-testid="statusTimeline">
       <ul className="statusTimelineUl">
         {statusConfig.map(
-          (config: any, index: React.Key | null | undefined) => {
+          (_config: any, index: React.Key | null | undefined) => {
+            const config = JSON.parse(JSON.stringify(_config));
```
Comment (amnambiar) [inline src/compositions/Timeline/Timeline.tsx]: I would suggest to remove this edit now, as the original intend was to modify the config with the calculated values. 

I couldn't test this bit, to ensure this copy-by-value (removing copy-by-reference) works as I started getting loading issues. Hence I suggest to not do this update here.

File: src/store/slices/testing.slice.ts [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+      coverageFile,
+      resultData,
+      runEnded,
+      fetching,
+      refetchMin,
+    };
+  } catch (e: any) {
+    console.log(e);
+    return thunkApi.rejectWithValue(e.response.data);
+  }
```
Comment (amnambiar) [inline src/store/slices/testing.slice.ts]: Isn't this too much of logic within slice. Is that alright?

File: src/store/slices/testing.slice.ts [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        name: form.name,
+        subject: form.subject,
+        owner, repo, githubToken
+      }
+    }));
+    await thunkApi.dispatch(clearAccessToken({}));
+
+    const response = await fetch<string>(thunkApi, {
+      method: 'POST', url: '/run', data: form.commitHash,
+      headers: {'Content-type': 'text/plain;charset=utf-8'}
```
Comment (amnambiar) [inline src/store/slices/testing.slice.ts]: This is the text/plain content-type that has been referred to above. Suggest to move to it an axios instance, for better maintenance

File: src/store/slices/testingHistory.slice.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+export const fetchHistory = createAsyncThunk("fetchHistory", async (payload: {}, thunkApi) => {
+  try {
+    const response = await fetch<Run[]>(thunkApi, { method: 'GET', url: '/run' });
+    return response.data;
+  } catch (e: any) {
+    return thunkApi.rejectWithValue(e.response.data);
+  }
+});
+
+export const updateRowStatus = createAsyncThunk("updateRowStatus", async (payload: { index: number, runId: string }, thunkApi) => {
```
Comment (amnambiar) [inline src/store/slices/testingHistory.slice.ts]: `updateRowStatus` could be renamed to `getRowStatus`, since it is not an update happening here

File: src/store/slices/testing.slice.ts (hunk 8 above)
Comment (amnambiar) [reply src/store/slices/testing.slice.ts]: My main concern is with doing computations using `status` and `state`, and such rendering specific logic within there. Please let me know your thoughts,

File: src/api/api.ts (hunk 1 above)
Comment (EehMauro) [reply src/api/api.ts:18]: This header is added in the action that needs it:
https://github.com/input-output-hk/dapps-certification-web/blob/288fc2954503c70ce27ef919d1dabf7a0b7ba51e/src/store/slices/testing.slice.ts#L84

Comment (EehMauro) [reply src/api/api.ts:18]: But, I'm adding a parameter in the fetch method so it can be included by passing an argument
(and it can be reused without repeating the header declaration)

File: src/api/api.ts (hunk 2 above)
Comment (EehMauro) [reply src/api/api.ts:55]: I'm passing the accessToken header in the action that needs it:
https://github.com/input-output-hk/dapps-certification-web/blob/288fc2954503c70ce27ef919d1dabf7a0b7ba51e/src/store/slices/repositoryAccess.slice.ts#L37-L38

Comment (EehMauro) [reply src/api/api.ts:55]: But also, I'm adding a parameter for this too

File: src/api/onRequest.ts (hunk 3 above)
Comment (EehMauro) [reply src/api/onRequest.ts:25]: Yes, we should improve all the error handling in PLT-8283

File: src/store/slices/testing.slice.ts (hunk 9 above)
Comment (EehMauro) [reply src/store/slices/testing.slice.ts]: Ok, now this is handled by a parameter

File: src/store/slices/testingHistory.slice.ts (hunk 10 above)
Comment (EehMauro) [reply src/store/slices/testingHistory.slice.ts]: Done!

File: src/store/slices/testing.slice.ts (hunk 8 above)
Comment (EehMauro) [reply src/store/slices/testing.slice.ts]: I agree, right now we have too much logic that can be simplified. But I think it'd be better to do that refactor in another PR.
In this task my only concern was to keep the component working without introducing too many changes to the internal logic.

File: src/compositions/Timeline/Timeline.tsx (hunk 7 above)
Comment (EehMauro) [reply src/compositions/Timeline/Timeline.tsx]: Ok! then I should move this update logic outside of the component rendering

File: src/components/LogsView/LogsView.css (hunk 6 above)
Comment (EehMauro) [reply src/components/LogsView/LogsView.css:11]: Ok! I'll fix it

File: src/components/CreateCertificate/CreateCertificate.tsx (hunk 5 above)
Comment (EehMauro) [reply src/components/CreateCertificate/CreateCertificate.tsx]: Ok, I'm working on it

File: src/components/CertificationMetadataForm/index.tsx (hunk 4 above)
Comment (EehMauro) [reply src/components/CertificationMetadataForm/index.tsx]: Ok, I'm working on this

Comment (EehMauro) [reply src/components/CertificationMetadataForm/index.tsx]: Ok, now is added again

File: src/components/LogsView/LogsView.css (hunk 6 above)
Comment (EehMauro) [reply src/components/LogsView/LogsView.css:11]: Done

File: src/components/CreateCertificate/CreateCertificate.tsx (hunk 5 above)
Comment (EehMauro) [reply src/components/CreateCertificate/CreateCertificate.tsx]: Done

File: src/compositions/Timeline/Timeline.tsx (hunk 7 above)
Comment (EehMauro) [reply src/compositions/Timeline/Timeline.tsx]: Done

File: src/compositions/Timeline/Timeline.tsx [hunk 11]
```diff
@@ -7,20 +7,12 @@ import "./Timeline.css";
...
-  const { buildInfo } = useAppSelector((state) => state.runTime);
 
   return (
     <div className="statusTimeline" data-testid="statusTimeline">
       <ul className="statusTimelineUl">
-        {statusConfig.map(
-          (config: any, index: React.Key | null | undefined) => {
-            if (buildInfo.runState === config.status) {
-              config.runTimeTaken = buildInfo.runTime
-            }
```
Comment (amnambiar) [inline src/compositions/Timeline/Timeline.tsx:19]: Why is this removed? 
@EehMauro We have to show the run time taken under each of the timeline phase once it is completed.

File: src/compositions/Timeline/Timeline.tsx (hunk 7 above)
Comment (amnambiar) [reply src/compositions/Timeline/Timeline.tsx]: The edit is not clear to me. I've added another comment though., pls check that

**Comment by amnambiar** [comment]
Found an issue -

Test InProgress state is retained now. But, in case of build failures, the run is not cleared and now there is no way to run a new test. 
https://github.com/input-output-hk/dapps-certification-web/assets/11058223/f5b3edf1-9574-4b84-8a0e-8214b42f5e8e

Can we show the test form, whenever it is not Running; i.e. both when finished/failure we don't want the previous run state in Testing, and instead the new test form. User can maybe go to History to see what has happened.

I believe, this will be the case/workflow when we try to read the state of any run InProgress within the profile from BE as well.

File: src/compositions/Timeline/Timeline.tsx (hunk 11 above)
Comment (EehMauro) [reply src/compositions/Timeline/Timeline.tsx:19]: Ok, is fixed now

**Comment by amnambiar** [comment]
1. Cannot still see runTime under the timeline states ([link](https://github.com/input-output-hk/dapps-certification-web/pull/45/files/be03de3fe648534d883b43e6bad0dff378e6527e#r1396719340))
2. When a testing is failed; e.g. Ali-Hill/plutus-apps --> 5f2c8c538 ; our workflow is still not clear. Clicking on 'Testing' stays in the failed build run. Since the states are maintained properly now here, this stays as a blocker for the user.
3. In this above run failure case, 'Abort' button is visible. In that case, we might not need to show the 'Abort' button as the test has already ended.

Above three pointers are still open in this PR. These can be anyhow resolved in Bug Fixes. Hence approving the PR for now.

Cc: @EehMauro @RSoulatIOHK

**Review by amnambiar — APPROVED** [review]

**Comment by RSoulatIOHK** [comment]
Move the remaining "bugs" to new tickets.
I think:
1) Should be fixed. I found that the times might be calculated in a strange way so I'd like to take the opportunity to see the triggers for "beginning" and "end" of each phase and maybe fix them.
2) When the build or a testing fails, it should give the opportunity to test again or test a new DApp or check the full report (if one exists).
3) Abort should be removed. A failed test campaign (from a build or a test) should be considered as finished.
